src/Json.php

```diff
@@ -70,6 +71,20 @@ public function retrieveValue(?array $data, ?ReflectionNamedType $type = null)
             $data = $data[$pathBit];
         }
 
+        if ($type instanceof ReflectionUnionType) {
+            $typesByType = $type->getTypes();
+            foreach ($type->getTypes() as $type) {
+                $typesByType[$type->getName()] = $type;
+            }
+            $type = match (gettype($data)) {
```

Comment:
If we're going to overwrite `$type`, this needs to be done after the checks for `$type->allowsNull()`.

If the type is `string|int|null` for example, the `ReflectionUnionType` will have `->allowsNull()` true, but the `string` one will not.